### ISSUE TITLE
Timeline maybe Start but without end

### ISSUE THREAD

**Issue by anpark** [uber/horovod#252]
https://github.com/uber/horovod/blob/f4c326ca25a142d1d2a117a87809fa7655724679/horovod/common/operations.cc#L936

when using nccl, thread not join, maybe timeline start but  without end, cause timeline next time start conflict

**Comment by alsrgv** [comment]
@anpark, do you actually see this overlap in timelines?

From the code just above:
```c++
        CUDA_CHECK(entries, "cudaEventSynchronize",
                   cudaEventSynchronize(done_event))
        for (auto it = entries.begin(); it != entries.end(); it++) {
          timeline.End(it->tensor_name, it->output);
          it->callback(Status::OK());
        }
        RELEASE_EVENT(entries, done_event);
```

Essentially, it will emit "End" event for a tensor before telling TensorFlow that tensor has finished the reduction.  In normal SGD situations, the tensor will not be reduced again while previous reduce operation is still not finished.

**Comment by anpark** [comment]
@alsrgv Yes, it's easy to be reappeared, i found that if detach, the same name tensor to be reduced maybe arrive before the "End" called.

**Comment by alsrgv** [comment]
@anpark, can you share the timeline file with this issue?

**Comment by anpark** [comment]
the timeline file is ok, because it's thread lock.

i have found the bug and fixed it:  tensor_states_ not thread safe, so thread lock should be put in timeline.Start, End, ActivityStart....and so on

by timeline i found two big performance problem:
1. new/delete for large buffer not efficiency, for example allgather, 300M buffer new/delete will cost much time, so should use buffer pool
2. mpi_allgather should change  to ncclallgather, in my test, mpi_allgather too slow for large buffer

**Comment by alsrgv** [comment]
@anpark, great catch!  Submitted a PR for the fix (#434).
